File: dotnet-mvc/CleanArchitecture/UseCases/SearchProductsUseCase.cs
using CleanArchitecture.Entities;
using CleanArchitecture.UseCases.Ports;

namespace CleanArchitecture.UseCases;

/// <summary>
/// Use Case: Search products by query string.
/// This belongs to the USE CASES layer (application business rules).
/// </summary>
public class SearchProductsUseCase
{
    private readonly IProductGateway _gateway;

    public SearchProductsUseCase(IProductGateway gateway)
    {
        _gateway = gateway;
    }

    public async Task<IEnumerable<Product>> Execute(string query)
    {
        if (string.IsNullOrWhiteSpace(query))
            return Enumerable.Empty<Product>();

        return await _gateway.Search(query);
    }
}

File: nodejs-mvc/clean-architecture/src/usecases/SearchProductsUseCase.ts
import { Product } from '../entities/Product';
import { ProductGateway } from './interfaces/ProductGateway';
import { ProductOutput, SearchInput } from './dto/ProductDTO';

/**
 * @layer Use Cases
 * @description Searches products by criteria.
 */
export class SearchProductsUseCase {
  constructor(private readonly gateway: ProductGateway) {}

  async execute(input: SearchInput): Promise<ProductOutput[]> {
    const products = await this.gateway.search(
      input.query,
      input.category,
      input.minPrice,
      input.maxPrice
    );
    return products.map(this.toOutput);
  }

  private toOutput(product: Product): ProductOutput {
    return {
      id: product.id,
      name: product.name,
      description: product.description,
      price: product.price,
      category: product.category,
      stockQuantity: product.stockQuantity,
      sku: product.sku,
      active: product.active,
      createdAt: product.createdAt.toISOString(),
      updatedAt: product.updatedAt.toISOString(),
    };
  }
}
